### PR TITLE
`Printf` maximum output size increased to 1Mb.

### PR DESCRIPTION
**Pull request by mzhurovich**
I didn't have a chance to test on Windows :(

### CODE REVIEW

File: Bricks/strings/printf.h [hunk 1]
```diff
@@ -37,26 +38,46 @@ namespace strings {
...
+  }
 #else
-  _vsnprintf_s(buf, max_formatted_output_length + 1, _TRUNCATE, fmt, ap);
-#endif
+  va_start(ap, fmt);
+  const int res = _vsnprintf_s(buffer, max_string_length_for_static_buffer + 1, max_string_length_for_static_buffer, fmt, ap);
   va_end(ap);
+  if (errno == ERANGE) {
+    const int large_buffer_length = std::min(_vscprintf(fmt, ap) + 1, max_formatted_output_length);
+    char large_buffer[new_buffer_length];
```
Comment (sompylasar) [inline Bricks/strings/printf.h]: `new_buffer_length` looks undefined, is it?

File: Bricks/strings/printf.h [hunk 2]
```diff
@@ -37,26 +38,46 @@ namespace strings {
 __attribute__((__format__(__printf__, 1, 2)))
 #endif
 inline std::string Printf(const char *fmt, ...) {
-  // `Printf()` crops the output to five KiB.
-  const int max_formatted_output_length = 5 * 1024;
+  // Most of the platforms now support thread locals, so 64Kb buffer seems reasonable.
+  const int max_string_length_for_static_buffer = 64 * 1024 - 1;
```
Comment (dkorolev) [inline Bricks/strings/printf.h]: `constexpr` here and below?

File: Bricks/strings/printf.h [hunk 3]
```diff
@@ -37,26 +38,46 @@ namespace strings {
...
   va_list ap;
-  va_start(ap, fmt);
 #ifndef CURRENT_WINDOWS
-  vsnprintf(buf, max_formatted_output_length + 1, fmt, ap);
+  va_start(ap, fmt);
+  const int res = vsnprintf(buffer, max_string_length_for_static_buffer + 1, fmt, ap);
+  va_end(ap);
+  if (res > max_string_length_for_static_buffer) {
+    const int large_buffer_length = std::min(res + 1, max_formatted_output_length + 1);
+    char large_buffer[large_buffer_length];
```
Comment (dkorolev) [inline Bricks/strings/printf.h]: Not: I believe non-`constexpr` local array size is a non-standard extension. Perhaps `vector<char>` here?

File: Bricks/strings/printf.h (hunk 1 above)
Comment (dkorolev) [reply Bricks/strings/printf.h]: +1

File: Bricks/strings/test.cc [hunk 4]
```diff
@@ -63,7 +63,12 @@ using current::strings::is_string_type;
 
 TEST(StringPrintf, SmokeTest) {
   EXPECT_EQ("Test: 42, 'Hello', 0000ABBA", Printf("Test: %d, '%s', %08X", 42, "Hello", 0xabba));
-  EXPECT_EQ(1024u * 5, Printf("%s", std::string(10000, 'A').c_str()).length());
+  const std::string string_60k_chars(60000, 'A');
+  EXPECT_EQ(60000u, Printf("%s", string_60k_chars.c_str()).length());
+  EXPECT_EQ(120000u, Printf("%s%s", string_60k_chars.c_str(), string_60k_chars.c_str()).length());
+  const std::string string_1m_chars(1000000, 'B');
+  EXPECT_EQ(1000000u, Printf("%s", string_1m_chars.c_str()).length());
+  EXPECT_EQ(1024u * 1024u - 1, Printf("%s%s", string_1m_chars.c_str(), string_1m_chars.c_str()).length());
```
Comment (dkorolev) [inline Bricks/strings/test.cc]: Excellent test case!

**Comment by mzhurovich** [comment]
@dkorolev @sompylasar PTAL

File: Bricks/strings/printf.h (hunk 3 above)
Comment (mzhurovich) [reply Bricks/strings/printf.h]: You're right, it's C99 and didn't go to C++11.

**Comment by dkorolev** [comment]
@mzhurovich -- yes, `&s[0]` should be good since C++11 ([link](https://stackoverflow.com/posts/33124544/revisions)).

Agree to keep it safe here though, perhaps with a `TODO(dkorolev)` to revisit as move move to the next Standard.

Thanks,
Dima

**Pull request merged**